tests/kernel/smp/src/main.c

```diff
@@ -652,23 +654,32 @@ ZTEST(smp, test_smp_ipi)
 	ztest_test_skip();
 #endif
 
-	TC_PRINT("cpu num=%d", arch_num_cpus());
+	int num_cpus = arch_num_cpus();
+	int num_ipis;
 
 	for (int i = 0; i < 3 ; i++) {
+		k_sched_lock();
+
 		/* issue a sched ipi to tell other CPU to run thread */
-		sched_ipi_has_called = 0;
+		atomic_clear(&sched_ipi_has_called);
+		atomic_clear(sched_ipi_on_cpu);
 		arch_sched_ipi();
 
-		/* Need to wait longer than we think, loaded CI
-		 * systems need to wait for host scheduling to run the
-		 * other CPU's thread.
-		 */
-		k_msleep(100);
+		k_busy_wait(DELAY_US);
```

Comment:
This doesn't look right.  The sense of that comment was that in CI environments[1],  IPI delivery is going to depend on scheduling on the host processors, which needless to say aren't going to be able to deliver within a microsecond-scale delay.  (FWIW, that comment sounds like my writing, but I have zero memory of this particular issue.  I don't think I wrote the IPI test originally though.)

Also k_busy_wait() tends to bound on local interrupt locks on most architectures and seems likely to make interrupt latency tests like this **worse** and not better.  I guess I don't see the value here?

[1]  Where the host is loaded by twister and SMP cores, which run in separate host threads, are going to see "time warp" effects relative to each other as they get scheduled in and out in competition with other build/test processes.

Reply:
On the contrary I think it is the original comment that is wrong. K_msleep() does trigger extra scheduler IPIs in addition to the explicit one here. Furthermore, this test does use this k_busy_wait() all over the place to give other CPUs the time to schedule and execute newly created threads, etc. So if this was a problem the entire test would be broken.
As it is, this IPI test is a bit pointless. We want to help new architecture implementors to do it right i.e. raise an IRQ on all CPUs but the current one and that is not tested anywhere. Given that the modification to the test do  work on all supported architectures already makes me wonder why you appear to be so concerned?